add maintainer docs for building and publishing (#2148)
ansible/ansible-documentation: MAINTAINERS.md

Update the list of active branches in the `hacking/tagger/tag.py` script on the `devel` branch.
Add the new stable branch and remove the lowest version from the `DEFAULT_ACTIVE_BRANCHES` tuple.

## Updating the package doc builds repository

The `ansible-community/package-doc-builds` repository holds the deployed output of package doc builds.
For each version of the Ansible package that is currently supported, there needs to be a corresponding branch in the `ansible-community/package-doc-builds` repository.

Before the docs for new package version can be built on Read the Docs, someone needs to create a branch as follows:

```bash
# Make sure your checkout is up to date.
git fetch upstream

# Create the new package version branch against the devel branch.
git checkout -b 11 upstream/devel

# Push the new package version branch to the repository.
git push upstream 11
```

## Updating the package docs build workflow

The build workflow lists the Ansible package versions as input options, for example:

```yaml
options:
- devel
- '11'
- '10'
- '9'
```

> After creating a new branch for the Ansible package version in the `ansible-community/package-doc-builds` repository, you should check that the build workflow lists that version as an option.

Reviewer: ```suggestion
After creating a new branch for the Ansible package version in the `ansible-community/package-doc-builds` repository, you should check that the `.github/workflows/build-package-docs.yaml` lists that version as an option.
```

Author: I don't think we need to be specific about the name of the workflow file here. That seems to interrupt readability imo.

If the version is not listed, open a PR against the `devel` branch to update the workflow file.

## Maintaining Read the Docs projects

There are two Read the Docs projects associated with the `ansible-documentation` repository, as follows:

* Ansible package
* Ansible core

### Configuration files

The Ansible package and Ansible core projects use separate configuration files, as follows:

* Ansible package configuration is the `package-doc-builds` repository at [https://github.com/ansible-community/package-doc-builds/blob/devel/.readthedocs.yaml](https://github.com/ansible-community/package-doc-builds/blob/devel/.readthedocs.yaml)
* Ansible core configuration is in the `ansible-documentation` repository at [https://github.com/ansible/ansible-documentation/blob/devel/.readthedocs.yaml](https://github.com/ansible/ansible-documentation/blob/devel/.readthedocs.yaml)

> Each branch has its own Read the Docs configuration file.

Most of the time there is no need to update the Read the Docs configuration files.
The only change that should be necessary between versions is the Python version, for example:

```yaml
build:

>   os: ubuntu-22.04

Reviewer: shouldn't we also need to update the ubuntu version now and then?

Author: yeah, I suppose we could also just change this to `ubuntu-latest` and not worry about it. fair point though. I think latest will be fine for these RTD configs because they are quite minimal and use only standard shell commands.

Author: my bad I guess. this was already `ubuntu-lts-latest`.

  tools:
    python: >-
      3.11
```

This Python version should match whatever version is used in the `ansible-documentation` repository to compile requirements and build the documentation.

> For example, the `stable-2.18` branch uses Python 3.11.

Reviewer: Is this the line that determines the python version - https://github.com/ansible/ansible-documentation/blob/devel/noxfile.py#L157
If so, can we add a link or 'see the noxfile.py' statement or something so we (aka me) remember where to look?

Author: Sounds good!

> The Python version should be also be 3.11 in the Read the Docs configuration for both Ansible package and Ansible core.

Reviewer: ```suggestion
The Python version should be also be 3.11 in the Read the Docs configuration (`.readthedocs.yaml`)  for both Ansible package and Ansible core.
```

Author: I'm not sure this adds that much. Since it's in the maintainers file, I'd be inclined to say the reader will know the actual file name.

To update the Ansible package project, do the following:

1. Ensure you have maintainer access to the project.
2. Navigate to the [package-doc-builds dashboard](https://app.readthedocs.org/projects/package-doc-builds/).
3. Select the **Add version** button and select the new version to activate, which should be a new `XX` branch.
4. Wait for the new version build to complete and then edit the main project settings.
5. From the **Default branch** drop-down, select the new version that should correspond to "latest" and save your changes.

To update the Ansible core project, do the following:

1. Ensure you have maintainer access to the project.
2. Navigate to the [ansible-core dashboard](https://app.readthedocs.org/projects/ansible-core/).
3. Select the **Add version** button and select the new version to activate, which should be a new `stable-*` branch.

### Hiding versions on Read the Docs

Hiding older versions removes them from the fly-out menu on Read the Docs as well as search results.
Older versions of the documentation that are not supported or EOL should be hidden.
It is also possible to hide doc builds before they are released to evaluate the content.

> At release day, all that is needed is to toggle the switch so the build is no longer hidden and, in the case of the package docs, update the default branch for the project.

Reviewer: So once Ansible 9 EOLs, the version flyout for package docs should only have latest and devel.

Author: TBH I'm not sure how that comment is related to this section. However, I'm afraid that's not quite the case. The version flyout will show `latest|devel|10|11` and any other active versions.

To hide versions, do the following:

1. Open the settings for the version you want to remove.
2. Select the **Hidden** option.

See [How to hide a version and keep its documentation online](https://docs.readthedocs.io/en/stable/guides/hiding-a-version.html) for more information.

## Building and deploying Ansible community documentation

> Content available at `https://ansible.readthedocs.io/projects/ansible/latest/` is built and deployed from the `ansible-documentation` repository using a GitHub workflow.

Reviewer 2: ```suggestion
Content available at https://ansible.readthedocs.io/projects/ansible/latest/ is built and deployed from the `ansible-documentation` repository using a GitHub workflow.
```

> Refer to the comments in the workflow itself for more information.

### Building community documentation

The Ansible package docs build workflow allows you to run a Sphinx build with the `make webdocs` target on GitHub hosted runners.
The workflow can build from any account or org as well as from the `ansible-documentation` repository.
For instance, if you wanted to build the content of a branch in someone's fork, you can provide those details as inputs when running the workflow.

To build the docs, do the following:

1. Navigate to the [Ansible package docs build](https://github.com/ansible/ansible-documentation/actions/workflows/build-package-docs.yaml) workflow.
2. Select the **Run workflow** drop-down to open the dialog.
3. Provide the necessary inputs and then select **Run workflow**.

> If the workflow fails for any reason, the `docs-bot` posts a message in the `#docs:ansible.com` channel on Matrix.

When the workflow run is complete, you can find the output available as an artifact on the **Summary** page.
The artifact is named **package-docs-build** and contains a tarball with the generated HTML and other build assets.

> 

Reviewer: Should we add something here about open the tarball locally and verify before approving the publish.

Author: Sure, I'll update with that.

Author: I put this detail under the **Reviewing deployments** section as an extra step.
